video/clip.go

```diff
@@ -0,0 +1,79 @@
+package video
+
+import (
+	"fmt"
+	"github.com/grafov/m3u8"
+)
+
+// Function to calculate the total duration of the HLS manifest
+func calculateTotalDuration(manifest *m3u8.MediaPlaylist) float64 {
+	if manifest == nil {
+		return 0.0
+	}
+
+	totalDuration := 0.0
+	for _, segment := range manifest.Segments {
+		if segment == nil {
+			break
+		}
+		totalDuration += segment.Duration
+	}
+	return totalDuration
+}
+
+// Checks if a given segment already exists in a slice of MediaSegments. Mainly used to ensure
+// we don't add duplicate segments when the playhead start/stop time falls in the same segment.
+func segmentExists(relSegs []*m3u8.MediaSegment, seg *m3u8.MediaSegment) bool {
+	for _, s := range relSegs {
+		if s.SeqId == seg.SeqId {
+			return true
+		}
+	}
+	return false
+}
+
+// Finds the segment in an HLS manifest that contains the timestamp (aka playhead) specified.
+func getRelevantSegment(allSegments []*m3u8.MediaSegment, relevantSegments *[]*m3u8.MediaSegment, playHeadTime float64, startingIdx uint64) (segNum uint64) {
```

Comment:
Why is relevantSegments an input and not an output?

Reply:
I think you can do it all with indexes at the point, so something like `getTimestampSegmentIndex(allSegments []*m3u8.MediaSegment, playHeadTime float64) (int, error)`

Reply:
But maybe it would make more sense if `getRelevantSegment` just returned an index, then in `ClipManifest` it can form the `relevantSegments` slice when it knows both the start and end index.

Reply:
Alright, updated code to use just indices instead of passing pointers around.